extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/MemoryCache.java

```diff
@@ -0,0 +1,130 @@
+package io.quarkus.oidc.runtime;
+
+import java.time.Duration;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+
+public class MemoryCache<T> {
+    private volatile Long timerId = null;
+
+    private Map<String, CacheEntry<T>> cacheMap = new ConcurrentHashMap<>();;
+    private AtomicInteger size = new AtomicInteger();
+    private final Duration cacheTimeToLive;
+    private final int cacheSize;
+
+    public MemoryCache(Vertx vertx, Optional<Duration> cleanUpTimerInterval,
+            Duration cacheTimeToLive, int cacheSize) {
+        this.cacheTimeToLive = cacheTimeToLive;
+        this.cacheSize = cacheSize;
+        init(vertx, cleanUpTimerInterval);
+    }
+
+    private void init(Vertx vertx, Optional<Duration> cleanUpTimerInterval) {
+        cacheMap = new ConcurrentHashMap<>();
+        if (cleanUpTimerInterval.isPresent()) {
+            timerId = vertx.setPeriodic(cleanUpTimerInterval.get().toMillis(), new Handler<Long>() {
+                @Override
+                public void handle(Long event) {
+                    // Remove all the entries which have expired
+                    removeInvalidEntries();
+                }
+            });
+        }
+    }
+
+    public void add(String key, T result) {
+        if (cacheSize > 0) {
+            if (!prepareSpaceForNewCacheEntry()) {
+                clearCache();
+            }
+            cacheMap.put(key, new CacheEntry<T>(result));
+        }
+    }
+
+    public T remove(String key) {
+        CacheEntry<T> entry = removeCacheEntry(key);
+        return entry == null ? null : entry.result;
+    }
+
+    public T get(String key) {
+        CacheEntry<T> entry = cacheMap.get(key);
+        return entry == null ? null : entry.result;
+    }
+
+    public boolean containsKey(String key) {
+        return cacheMap.containsKey(key);
+    }
+
+    private void removeInvalidEntries() {
+        long now = now();
+        for (Iterator<Map.Entry<String, CacheEntry<T>>> it = cacheMap.entrySet().iterator(); it.hasNext();) {
```

Comment:
`entrySet` javadoc says _If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation, or through the setValue operation on a map entry returned by the iterator) the results of the iteration are undefined._ I wonder what happens when you iterate here and entry is added or removed, especially the same one as you iterate over here.

Reply:
@michalvavrik, I suppose the worst what can happen is that the cache entry which is actually there won't be found which will simply cause another remote call with that entry eventually overridden. Like I said this is not a new code, it has been used for a long while in the UserInfo cache support,  I don't mind improving it but would prefer not to focus on it much here, this PR only makes this code reusable.  I think we'll want to support Redis cache for prod cases.

Reply:
Okay, I didn't mention this part of code didn't change. I was worried about `next` variable on the next line being null, but in this light it's fine. Let's keep it.

Reply:
@michalvavrik 

>I was worried about next variable on the next line being null or delete operation throwing an exception

Sure, let me add a couple of protections

Reply:
@michalvavrik How does it look now ? I've added NPE check for `next` and ISE catch block for `remove`

Reply:
lgtm, thanks